### PR TITLE
week6 - tree-related problems

### PR DESCRIPTION
**Pull request by maknoon**
*isSubtree*
- `$ make st && ./st`
- TEBOW (38min for 2 different approaches), full implementation of one approach took additional 30min w/ debugging. solution breaks the process down into two recursive functions: `isSameTree` validates if two trees are structurally & content-wise the same, and the main function `isSubTree` traverses the parent tree to find the first node in it that is a possible match to the root of the subtree we are validating. tests are 50% done, although I did write down the equivalence classes in the comments
- the other approach I had in mind, which I think could save space, is to pre-order traverse S and store its contents in a string or array. then do a similar traversal of T as before looking for the first possible match & then pre-order traversing that node as well. finally, check if both strings/arrays match. in this way we can minimize space but also check if the trees are structurally equivalent. may try to implement this later  

- `TODO` finish up tests, maybe write a helper to construct the strings from an array, also do the challenge problems

### CODE REVIEW

File: week6/isSubtree.cc [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ * bool isSubtree(tNode *s, tNode *t)
+ * inputs: *t - pointer to the root of the subtree we are validating
+           *s - pointer to the root of the tree we are using as reference
+ * output: true if t is a subtree of s, false otherwise
+ * notes: main solution function
+ */
+bool isSubtree(tNode *t, tNode *s)
+{
+  // base case + null cases
+  if (!s && !t) return true;
```
Comment (loresfca) [inline week6/isSubtree.cc]: just a tiny correction here: if you have t as null, regardless if s is null, t will be a subtree so it will always return true

**Review by loresfca — COMMENTED** [review]
Hey! you did a marvelous job!  just added a comment about the base cases but I think we got to a very similar solution :) Great job with the comments and modularizing your code

File: week6/isSubtree.cc [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+bool isSubtree(tNode *s, tNode *t);
+
+//== FUNCTIONS
+/*
+ * void deleteTree(tNode*& root)
+ * inputs: root - root node of the tree that we want to delete
+ * output: none
+ * notes: post-order traversal helper. don't need those memory leaks :P
+ */
+void deleteTree(tNode*& root)
```
Comment (charientism) [inline week6/isSubtree.cc:127]: :)

File: week6/isSubtree.cc [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  // TEST 7 - complex fail w/ duplicate numbers
+  cout << "TEST 1: t={2,null,2,10,3}, s={2,null,3}" << endl;
+  // TODO
+}
+
+int main(void)
+{
+  tests();
+  return 0;
+}
```
Comment (charientism) [inline week6/isSubtree.cc:274]: Nice job with your written TEBOW IT notes, particularly optimize! You're right that the substring of another string approach would decrease your time complexity but increase your space complexity. Calling out that kind of trade-off and getting your interviewer's buy-in is exactly the way you'd want to handle this is an interview!

File: week6/isSubtree.cc [hunk 4]
```diff
@@ -0,0 +1,160 @@
+/* isSubtree.cc
+ * author: connie
+ *
+ * IS SUBTREE
+ * Checks if a binary tree T is a subtree of binary tree S.
+ * A subtree is defined as have the same structure as another tree's root
+ * or child nodes -
+ * i.e., T does not necessarily have to be a node actually in tree S.
```
Comment (charientism) [inline week6/isSubtree.cc:8]: 👍

File: week6/isSubtree.cc [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  else {
+    // only check isSameTree if the node data is the same
+    if ((s->data == t->data) && isSameTree(s,t)) return true;
+
+    // otherwise:
+    // recurse on L and R children of s to find another suitable node
+    // as long as either side comes up with a matching subtree, we are done
+    else return (isSubtree(t, s->left) || isSubtree(t, s->right));
+  }
+}
```
Comment (charientism) [inline week6/isSubtree.cc:185]: Great job on the implementation! Your work is always extremely readable (well-structured and nicely commented) which I very much appreciate :)

**Review by charientism — APPROVED** [review]

File: week6/isSubtree.cc (hunk 1 above)
Comment (maknoon) [reply week6/isSubtree.cc]: nice! thanks for tip :D fixed

**Pull request merged**